provisioner/kuberpak/internal/util/crd.go

```diff
@@ -0,0 +1,192 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func CreateOrUpdateCRD(ctx context.Context, cl client.Client, crd *apiextensionsv1.CustomResourceDefinition) (controllerutil.OperationResult, error) {
+	createCRD := crd.DeepCopy()
+	createErr := cl.Create(ctx, createCRD)
+	if createErr == nil {
+		return controllerutil.OperationResultCreated, nil
+	}
+	if !apierrors.IsAlreadyExists(createErr) {
+		return controllerutil.OperationResultNone, createErr
+	}
+	if updateErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentCRD := &apiextensionsv1.CustomResourceDefinition{}
+		if err := cl.Get(ctx, client.ObjectKeyFromObject(crd), currentCRD); err != nil {
+			return err
+		}
+		crd.SetResourceVersion(currentCRD.GetResourceVersion())
+
+		if err := validateCRDCompatibility(ctx, cl, currentCRD, crd); err != nil {
+			return fmt.Errorf("error validating existing CRs against new CRD's schema for %q: %w", crd.Name, err)
+		}
+
+		// check to see if stored versions changed and whether the upgrade could cause potential data loss
+		safe, err := safeStorageVersionUpgrade(currentCRD, crd)
+		if !safe {
+			return fmt.Errorf("risk of data loss updating %q: %w", crd.Name, err)
+		}
+		if err != nil {
+			return fmt.Errorf("checking CRD for potential data loss updating %q: %w", crd.Name, err)
+		}
+
+		// Update CRD to new version
+		if err := cl.Update(ctx, crd); err != nil {
+			return fmt.Errorf("error updating CRD %q: %w", crd.Name, err)
+		}
+		return nil
+	}); updateErr != nil {
+		return controllerutil.OperationResultNone, updateErr
+	}
+	return controllerutil.OperationResultUpdated, nil
+}
+
+func keys(m map[string]apiextensionsv1.CustomResourceDefinitionVersion) sets.String {
+	return sets.StringKeySet(m)
+}
+
+func validateCRDCompatibility(ctx context.Context, cl client.Client, oldCRD *apiextensionsv1.CustomResourceDefinition, newCRD *apiextensionsv1.CustomResourceDefinition) error {
```

Comment:
> This may also be a good candidate to be shared
👍 Just because its shared doesn't mean other provisioners have to use it. 

For now, I'd suggest moving most of this into an `internal` package tree at the repo root so other areas of the repo can use it. I'd hold back on actually exporting any Go APIs until/unless we'd be comfortable telling others they're ready to be consumed. That reminds me, have we defined a versioning strategy yet? If not, maybe we need an issue for that?

Reply:
Yep I've been playing around moving the nested internal and cmd directories to the root directory here: https://github.com/timflannagan/rukpak/pull/2. Maybe I'm playing this a bit loose right now, but I think it's fine to export functions/methods within an internal directory, at least for the time being while we flesh out provisioner implementations. We can always cleanup the library once we get a better feel for what abstractions are needed over time. Thoughts?